Makefile

```diff
@@ -114,14 +115,11 @@ vs_cmake:
 		-DCMAKE_TOOLCHAIN_FILE=$(VCPKG_WIN_PATH) \
 		-DKOMPUTE_EXTRA_CXX_FLAGS=$(VS_KOMPUTE_EXTRA_CXX_FLAGS) \
 		-DCMAKE_INSTALL_PREFIX=$(VS_INSTALL_PATH) \
-		-DKOMPUTE_OPT_INSTALL=1 \
```

Comment:
I'm now testing the build in windows - did you manage to get it working? I updated this make command to use the repo build:

```
vs_cmake:
	$(CMAKE_BIN) \
		-Bbuild \
		$(VS_CMAKE_EXTRA_FLAGS) \
		-DCMAKE_CXX_FLAGS=$(VS_KOMPUTE_EXTRA_CXX_FLAGS) \
		-DCMAKE_BUILD_TYPE=$(VS_BUILD_TYPE) \
		-DCMAKE_INSTALL_PREFIX=$(VS_INSTALL_PATH) \
		-DKOMPUTE_OPT_INSTALL=ON \
		-DKOMPUTE_OPT_BUILD_TESTS=ON \
		-DKOMPUTE_OPT_BUILD_SHADERS=ON \
		-DKOMPUTE_OPT_CODE_COVERAGE=OFF \
		-DKOMPUTE_OPT_BUILD_DOCS=OFF \
		-G "Visual Studio 16 2019"
```

And I get the following error:

```
-- Running vulkaninfo was successful. Parsing the output...
-- GPU0 supports Vulkan API version '1.2.204'.
Valid GPU (GPU0) for Vulkan header version 1.2.203 found. GPU0 supports up to Vulkan 1.2.204.
-- Module support is disabled.
-- Version: 8.1.1
-- Build type: Debug
-- CXX_STANDARD: 14
-- Required features: cxx_variadic_templates
-- Setting KOMPUTE_OPT_LOG_LEVEL to according to the build type
CMake Error at C:/Users/axsau/Programming/lib/vcpkg/scripts/buildsystems/vcpkg.cmake:499 (_add_library):
  _add_library INTERFACE library requires no source arguments.
Call Stack (most recent call first):
  src/shaders/glsl/CMakeLists.txt:19 (add_library)


CMake Error at src/shaders/glsl/CMakeLists.txt:22 (target_include_directories):
  Cannot specify include directories for target "kp_shader" which is not
  built by this project.


CMake Error at C:/Users/axsau/Programming/lib/vcpkg/scripts/buildsystems/vcpkg.cmake:499 (_add_library):
  _add_library INTERFACE library requires no source arguments.
Call Stack (most recent call first):
  test/shaders/glsl/CMakeLists.txt:21 (add_library)


CMake Error at test/shaders/glsl/CMakeLists.txt:26 (target_include_directories):
  Cannot specify include directories for target "test_shaders_glsl" which is
  not built by this project.


-- Configuring incomplete, errors occurred!
See also "C:/Users/axsau/Programming/vulkan/vulkan-kompute/build/CMakeFiles/CMakeOutput.log".
See also "C:/Users/axsau/Programming/vulkan/vulkan-kompute/build/CMakeFiles/CMakeError.log".
make: *** [Makefile:112: vs_cmake] Error 1
```

It seems to be complaining at a couple of things, including the add_library include here:

```
add_library(kp_shader INTERFACE "${CMAKE_CURRENT_BINARY_DIR}/ShaderOpMult.hpp"
    "${CMAKE_CURRENT_BINARY_DIR}/ShaderLogisticRegression.hpp")
```

Have you seen this before in your side as well?

Reply:
I'm also testing the android examples and it seems there are some strange examples happening 
![image](https://user-images.githubusercontent.com/1447507/181361248-77f1584e-c836-4acd-89d7-d3075986c373.png)

I do wonder if it is to do with the cmake version in windows, however it's cmake 1.18.0

Reply:
Ok interseting I was able to upgrade the version of my cmake to 3.20.4 and it seems that it was able to build CMAKE without complaining this time - I am suspecting we may need to increase the minimum required version of cmake...

Reply:
I am still running into some errors when building in windows. Before when compiling against subrepo builds I always ran into issues with gtest as that library is supposed to be compiled without multiprocessing runtime as otherwise it gives errors in windows. That was achieved by running the build with the `VS_KOMPUTE_EXTRA_CXX_FLAGS ?= "/MT"` flag, which although is extremely slow would result in successful builds. In the current case unfortunately it doesn't really fix this and the same error appears to when you run with the multiprocessing flag (ie `VS_KOMPUTE_EXTRA_CXX_FLAGS ?= "/MP"`). The error is basically a linking error:
```
gtest_maind.lib(gtest_main.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MTd_StaticDebug' doesn't
 match value 'MDd_DynamicDebug' in TestAsyncOperations.obj [C:\Users\axsau\Programming\vulkan\vulkan-kompute\build\test\ko
mpute_tests.vcxproj]
```

![image](https://user-images.githubusercontent.com/1447507/181438165-3c9d49b1-759a-404e-aec3-a52764f6a0dc.png)

I suspect there may be something in the build configuration that may be forcing the multiprocessing build somewhere - have you seen this on your side?

Reply:
Fixed it by setting `gtest_force_shared_crt` when compiling.
https://github.com/google/googletest/tree/main/googletest#visual-studio-dynamic-vs-static-runtimes

Reply:
> Fixed it by setting gtest_force_shared_crt when compiling.
https://github.com/google/googletest/tree/main/googletest#visual-studio-dynamic-vs-static-runtimes

Interesting that's quite useful to know as this has been bugging me for a while now. I was thinking it may be worth putting it behind a flag but to be honest if it's for tests mainly it does seem like providing too much flexibility for users doesn't make as  much sense unless we need it for development, so happy to leave it as it is. I can confirm that it's now building correctly.

Unfortunately it seems that the execution of the tests themselves are not performing as expected - I am currently seeing the tests just freeze on the `[ RUN      ] TestAsyncOperations.TestManagerAsyncExecution` step. It basically gets stuck on

```
[Jul 28 2022 09:35:17] [debug] [C:\Users\axsau\Programming\vulkan\vulkan-kompute\src\OpTensorSyncLocal.cpp:57] Kompute OpTensorSyncLocal preEval called
[Jul 28 2022 09:35:17] [debug] [C:\Users\axsau\Programming\vulkan\vulkan-kompute\src\OpAlgoDispatch.cpp:48] Kompute OpAlgoDispatch preEval called
[Jul 28 2022 09:35:17] [debug] [C:\Users\axsau\Programming\vulkan\vulkan-kompute\src\OpTensorSyncLocal.cpp:57] Kompute OpTensorSyncLocal preEval called
[Jul 28 2022 09:35:17] [debug] [C:\Users\axsau\Programming\vulkan\vulkan-kompute\src\Sequence.cpp:132] Kompute sequence submitting command buffer into compute queue
```

Is this something you managed to run successfully?

> Turns out OUTPUT for add_custom_command requires CMake 3.20. https://cmake.org/cmake/help/latest/command/add_custom_command.html
> 
> With that we need to update to CMake 3.20, since a lot of stuff won't work any more like I planed it to.
But this should be fine since even Debian uses CMake 3.23.2-1.


In regards to the cmake piece, I think for now we can go ahead with just making it minimum 3.20 - we can validate as followup if there is a need to reduce the version. We shoudl validate if ubuntu LTS users (and potentially for some 18.04 / 19.04) woudl still get cmake of at least cmake 3.20+

Reply:
Regarding the Async tests, this is something I can confirm. This happens for me as well. 
`TestManagerAsyncExecutionTimeout` fails and the other two never finish their execution. Does this also happen on origin/master (not my fork)? I'm unable to test it, since it fails to compile for me.

Reply:
I also fixed the `_add_library INTERFACE library requires no source arguments.` error on Windows.

Reply:
Ok thank you for confirming, ok I can see that master also replicates the freezing on the parallel test, I will have a deeper look today at why this is the case. Thanks for confirming from your side.

Reply:
Right now that I was running the tests on the windows side I did notice that the master branch does seems to run the full compile significantly faster, this was particularly the case on the non-multiprocessing build. Would you know why it woudl be significantly slow with this new build structure? We could have a look at optimization in a separate PR but wanted to ask just in case

Reply:
This is because they are not build in parallel. Should be fixed by appending `--parallel` at the end of the CMake command, like I have done in the Makefile for `vs` builds. But needs to be tested further.